### ISSUE TITLE
How could I speed up the similarity matrix computation?

### ISSUE THREAD

**Issue by JannikSchneider12** [matchms/matchms#569]
Hey,

This is not an actual issue regarding the project, but I didn't know where to ask instead. So I have a dataframe that contains m/z and intensity arrays as well as cluster labels. I want to evaluate the clustering by applying the silhouette score. For now my approach looks like this:

import numpy as np
from matchms import calculate_scores, Spectrum
from matchms.similarity import CosineGreedy
from sklearn.metrics import silhouette_score
from sklearn.preprocessing import MinMaxScaler
import logging

# Configure logging to suppress warnings
logging.getLogger("matchms").setLevel(logging.ERROR)

# Step 1: Convert DataFrame rows into matchms Spectra
def row_to_spectrum(row):
    return Spectrum(mz=np.array(row['m/z array']), intensities=np.array(row['intensity array']))

# Assuming DataFrame containing mass spectra data
your_dataframe = df_read

# Convert DataFrame rows into matchms Spectra
spectrums = [row_to_spectrum(row) for _, row in your_dataframe.iterrows()]

# Step 2: Calculate similarity scores between all pairs of spectra
scores = calculate_scores(spectrums, spectrums, CosineGreedy(), is_symmetric=True).to_array()

similarity_matrix = scores['CosineGreedy_score']

# Convert similarity scores to dissimilarity scores (1 - similarity)
dissimilarity_matrix = 1 - similarity_matrix

# Optionally scale the dissimilarity matrix to [0, 1]
scaler = MinMaxScaler()
dissimilarity_matrix_scaled = scaler.fit_transform(dissimilarity_matrix)

# Calculate silhouette score
silhouette_avg = silhouette_score(dissimilarity_matrix_scaled, labels=df_read['Cluster_Index'], metric='precomputed')

print(f"Silhouette score: {silhouette_avg:.2f}")

However even for small sizes like ~10000 spectra it tooks a lot of time and also requires a lot of RAM. So I wanted to ask if someone has an idea how I could speed up my calculations and/or use less RAM.

Thanks for your help and sorry if this is a bit out of the scope of an issue regarding matchms

**Comment by niekdejonge** [comment]
@JannikSchneider12 thanks for the clear issue and sorry for the delay in response, I think it was missed at first. 
I am not familiar with the silhouette score, but do you want a cosine score between all 10000 times 10000 spectra? If you want this it will be hard to circumvent, since the number of pairs you are calculating in this case is 10000 times 10000  = 100.000.000 similarity scores, which will result in a high ram usage and calculation time. 
In matchms some solutions are implemented to prevent this issue, by using sparse matrixes, here we only store the values that are needed, for the cosine score you could for instance add a requirement that the precursor mz is (almost) equal. In that case matchms will only store the similarity scores if the precursor mz is equal.  You can use the Pipeline for this, see the readme for more info on how to do this. Similarly you can filter the scores by setting "filter_by_range", {"name": "CosineGreedy_score", "low": 0.2} to only keep cosine scores above 0.2. Hope this helps, let me know if you need more info.

**Comment by tornikeo** [comment]
@JannikSchneider12  For greedy cosine, I solved this issue by writing a custom CUDA kernel with numba. It's around x100-x500 faster than matchms. You can use it online on [colab](https://colab.research.google.com/drive/15RpnFVeiJmpR7i3VGF8qtBpA_Fsdb2EQ?usp=sharing), just upload your own spectra files. The source code for the project can be accessed [here](https://github.com/tornikeo/cudams). Disclaimer: I'm the author of the CudaMS project :smile:

**Comment by niekdejonge** [comment]
That is great! We didn't have the time yet to build something like this into matchms. I had a quick look at your package and it seemed like it would integrate nicely with matchms. It might be nice to have a look at integrating this directly into matchms itself (if you are interested in that), or at least mentioning your package in tutorials, so people know that if they need high speeds it is available already. The only issue for integrating it into matchms directly, to me would be the torch package, since it is quite large and might be an annoying dependency for other packages in matchms. But it is possible to make it an optional package when installing matchms. @JannikSchneider12 Would you be interested in integrating your package more with matchms? @florian-huber Do you think that would be a good idea, considering the (optional) pytorch dependency.

**Comment by tornikeo** [comment]
@niekdejonge  It is possible to discard pytorch, it will just require me to add a new numba kernel that does sparse reduction. Pytorch is only really used in "sparse" mode, where I "threshold" the output scores above, say 0.75. This is necessary because my case is comparing 10k spectra with 1.5mln, and the full dense output size is around 180GB and most of it is near-zero. 

However, keep in mind that the main reason pytorch is so large is that it depends on CUDA. And we still need it, unfortunately. We can optionally swap out the torch requirement to CuPy instead,  but I don't know how much of a relief that'd be - pytorch is much more well documented and widely used than CuPy, and comes with some a large set of docker images too.

About integrating this into matchms - it will make matchms dependent on cudatoolkit and build-essential, since each run requires compiling a custom cuda kernel. It might make matchms package inaccessible for many non-technical users, as cudatoolkit is hard to install, but I'm open to suggestions here.

**Comment by niekdejonge** [comment]
Yes, removing Pytorch won't really help and using the sparse mode is very helpfull, since I think most people that require these kinds of speeds will run into the same memory issues as you. 

Before we had RDKit as an optional package when installing matchms, by running pip install matchms[RDKIT] and every function relying on RDKit would first check if rdkit was installed and otherwise give an informative error. So we could try implementing it so pytorch etc can be installed by running pip install matchms[GPU]. If that is doable with the Cuda and numba packages. We could than add a separate test pipeline that checks that these functions work when having these packages installed. 

However, from a maintenance standpoint, it might be easier to keep your package separated, but in that case we should promote your implementation a bit more so others easily find your package when running into this issue. I will think about it a bit more. @florian-huber and @hechth what do you think? 

For now, I will add your package to the Tools building on matchms list in the readme.

**Comment by florian-huber** [comment]
Sorry, coming back to this issue very late.
@julianpollmann and I want to start looking at speeding up similarity calculations again and there are some nice starting points here. However, pytorch (due to CUDA mostly) is really a large and rather complicated dependency. Still, given the promising speed up it would be a viable option to consider. But I would probably use what Niek mentioned a "GPU" or "CUDA" version of matchms with pytorch, and the "normal" package without.
